File: HomeWork/src/main/java/week03/Solution106.java
package week03;

import common.TreeNode;

public class Solution106 {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        this.inorder = inorder;
        this.postorder = postorder;
        return build(0, inorder.length - 1, 0, postorder.length - 1);

    }

    private TreeNode build(int il, int ir, int pl, int pr) {

        if (il > ir) {
            return null;
        }

        TreeNode root = new TreeNode(postorder[pr]);

        for (int i = il; i <= ir; i++) {
            if (inorder[i] == postorder[pr]) {
                TreeNode left = build(il, i - 1, pl, pl + i - il - 1);
                TreeNode right = build(i + 1, ir, pr - ir + i, pr - 1);

                root.left = left;
                root.right = right;
            }

        }
        return root;

    }

    private int[] inorder;
    private int[] postorder;

    public static void main(String[] args) {
        Solution106 s = new Solution106();
        // int[] inorder = { 9, 3, 15, 20, 7 };
        // int[] postorder = { 9, 15, 7, 20, 3 };
        int[] inorder = { 2, 1 };
        int[] postorder = { 2, 1 };
        s.buildTree(inorder, postorder);
    }
}

File: Recursion/src/main/python/solution106.py
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def build(inorder, postorder):
            if len(inorder) == 0:
                return None

            root = TreeNode(postorder[-1])

            for i in range(len(inorder)):
                if inorder[i] == postorder[-1]:
                    left = build(inorder=inorder[0:i], postorder=postorder[0:i])
                    right = build(
                        inorder=inorder[i + 1 :],
                        postorder=postorder[i : len(postorder) - 1],
                    )

                    root.left = left
                    root.right = right

            return root

        return build(inorder, postorder)
